### ISSUE TITLE
"dukpy is currently not production ready"

### ISSUE THREAD

**Issue by maximinus** [amol-/dukpy#32]
This is stated in the README.rst file.

Since *production ready* is likely a fairly subjective status, what steps would it take to remove this status?

The inclusion of this library in the new version of pypac - which my team uses - has led us to wonder how we could change this status.

**Comment by amol-** [comment]
The note is related to the fact that the codebase is mostly C and thus can have bugs that crash your whole Python interpreter.
Also I used it only during assets preparation phase of web applications (so at "compile time") so never had a DukPy app running longterm to check eventual leaks, nor I ever had a chance to test it on a real multithreaded environment to see how if race conditions or similar problems arise.

Until users or myself had a chance to test it in those three conditions I wouldn't consider DukPy safe for production use.

**Comment by viernullvier** [comment]
I've stumbled over this issue just today while trying to fix a bug that turned out to be unrelated to dukpy.

I'm happy to inform you that we've now been using dukpy in production for _over three years_. Here's the commit:

![Bildschirmfoto 2019-06-27 um 23 08 36](https://user-images.githubusercontent.com/1492179/60301566-03a3ed00-9932-11e9-8a02-222efcf158d7.png)

First, a bit of context: I'm the lead developer of [enter](https://enter.dcb.ug/en/), a fully-featured and highly customizable attendee management system. To achieve the "highly customizable" part, for each event we have to a) handle a completely individual set of data fields, including statistics and b) have full control over the registration flow, including server-side validation.

For a), we use dukpy to run both individually written map/reduce functions on our data to generate charts and export scripts to generate Excel/CSV reports per customer request. For b), we use dukpy to run event-specific hooks into the registration process, for example before displaying the registration page, during the server-side input validation and before actually storing any user input in the database. On top of that, we use it to perform bulk manipulation on large amounts of entries.

We run dukpy in separate threads in order to get at least basic sandboxing and timeouts. I couldn't notice any crashes or memory leaks yet. So far, it has never caused any trouble. It's just there, in its right place, working as reliably and efficiently so that we don't even notice it. I've been through multiple options before [see this StackOverflow question](https://stackoverflow.com/questions/27842069/i-need-to-run-untrusted-server-side-code-in-a-web-app-what-are-my-options), including PyExecJS and PyV8 which both kind of worked but either had performance issues when running on larger amounts of data or show strange behavior when converting between Python and JS objects.

In conclusion, I'd actually consider dukpy ready for production.
